File: cpp/src/DirectoryManager.cpp
#include "treon/DirectoryManager.h"
#include "treon/Constants.h"

#include <QStandardPaths>
#include <QDir>
#include <QDebug>

using namespace treon;

DirectoryManager& DirectoryManager::instance() {
    static DirectoryManager instance;
    return instance;
}

DirectoryManager::DirectoryManager(QObject* parent)
    : QObject(parent)
    , m_settings(new QSettings(this))
{
    qDebug() << "Initializing DirectoryManager";
}

QUrl DirectoryManager::getLastOpenedDirectory() const {
    const QString lastDir = m_settings->value(UserDefaultsKeys::lastOpenedDirectory).toString();
    
    if (!lastDir.isEmpty()) {
        QDir dir(lastDir);
        if (dir.exists()) {
            qDebug() << "Using last opened directory:" << lastDir;
            return QUrl::fromLocalFile(lastDir);
        } else {
            qDebug() << "Last opened directory no longer exists, falling back to Documents";
            m_settings->remove(UserDefaultsKeys::lastOpenedDirectory);
        }
    }
    
    // Fall back to Documents directory
    const QString documentsPath = QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation);
    const QUrl documentsUrl = QUrl::fromLocalFile(documentsPath);
    qDebug() << "Using fallback directory:" << documentsPath;
    return documentsUrl;
}

void DirectoryManager::saveLastOpenedDirectory(const QUrl& fileUrl) {
    const QUrl directoryUrl = QUrl::fromLocalFile(QFileInfo(fileUrl.toLocalFile()).absolutePath());
    const QString directoryPath = directoryUrl.toLocalFile();
    
    m_settings->setValue(UserDefaultsKeys::lastOpenedDirectory, directoryPath);
    qDebug() << "Saved last opened directory:" << directoryPath;
}

void DirectoryManager::clearLastOpenedDirectory() {
    m_settings->remove(UserDefaultsKeys::lastOpenedDirectory);
    qDebug() << "Cleared last opened directory";
}

File: Treon/Services/DirectoryManager.swift
//
//  DirectoryManager.swift
//  Treon
//
//  Created by Vivek on 2024-10-19.
//  Copyright © 2024 Treon. All rights reserved.
//

import Foundation
import os.log

/// Manages directory memory functionality for file dialogs
@MainActor
class DirectoryManager {
    static let shared = DirectoryManager()
    
    private let logger = Logger(subsystem: "com.treon.app", category: "DirectoryManager")
    
    private init() {
        logger.info("Initializing DirectoryManager")
    }
    
    // MARK: - Directory Memory
    
    /// Gets the last opened directory from UserDefaults
    /// Falls back to Documents directory if no valid directory is stored
    func getLastOpenedDirectory() -> URL {
        if let data = UserDefaults.standard.data(forKey: UserDefaultsKeys.lastOpenedDirectory),
           let url = URL(dataRepresentation: data, relativeTo: nil) {
            // Verify the directory still exists
            var isDirectory: ObjCBool = false
            if FileManager.default.fileExists(atPath: url.path, isDirectory: &isDirectory) && isDirectory.boolValue {
                logger.info("Using last opened directory: \(url.path)")
                return url
            } else {
                logger.info("Last opened directory no longer exists, falling back to Documents")
                // Clear the invalid directory
                UserDefaults.standard.removeObject(forKey: UserDefaultsKeys.lastOpenedDirectory)
            }
        }
        
        // Fall back to Documents directory
        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first ?? URL(fileURLWithPath: NSHomeDirectory())
        logger.info("Using fallback directory: \(documentsURL.path)")
        return documentsURL
    }
    
    /// Saves the directory of the given file URL to UserDefaults
    func saveLastOpenedDirectory(url: URL) {
        let directoryURL = url.deletingLastPathComponent()
        
        let data = directoryURL.dataRepresentation
        UserDefaults.standard.set(data, forKey: UserDefaultsKeys.lastOpenedDirectory)
        logger.info("Saved last opened directory: \(directoryURL.path)")
    }
    
    /// Clears the stored directory memory
    func clearLastOpenedDirectory() {
        UserDefaults.standard.removeObject(forKey: UserDefaultsKeys.lastOpenedDirectory)
        logger.info("Cleared last opened directory")
    }
}
